File: Codes/2116-check-if-a-parentheses-string-can-be-valid.go
/*
 * @Author: LetMeFly
 * @Date: 2025-03-23 10:54:53
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2025-03-23 10:56:04
 */
package main

func canBeValid(s string, locked string) bool {
    l, r := 0, 0
    for i := range s {
        if locked[i] == '0' {
            l--
            r++
            if l < 0 {
                l = 1
            }
        } else {
            if s[i] == '(' {
                l++
                r++
            } else {
                l--
                r--
                if r < 0 {
                    return false
                }
                if l < 0 {
                    l = 1
                }
            }
        }
    }
    return l == 0
}

File: Codes/2116-check-if-a-parentheses-string-can-be-valid.py
'''
Author: LetMeFly
Date: 2025-03-23 10:49:44
LastEditors: LetMeFly.xyz
LastEditTime: 2025-03-23 10:51:19
'''
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        l = r = 0
        for i in range(len(s)):
            if locked[i] == '0':
                l -= 1
                r += 1
                if l == -1:
                    l = 1
            else:
                if s[i] == '(':
                    l += 1
                    r += 1
                else:
                    l -= 1
                    r -= 1
                    if r < 0:
                        return False
                    if l < 0:
                        l = 1
        return not l